pkg/kv/kvserver/replica_raftlog_test.go

```diff
@@ -0,0 +1,303 @@
+	if ln := len(app.Entries); ln > 0 {
+		rt.mu.readable = app.Entries[ln-1].Index
+	}
+}
+
+func (rt *replicaLogStorageTest) truncate(
+	ctx context.Context, prev, next kvserverpb.RaftTruncatedState,
+) {
+	if next.Index == prev.Index {
+		return
+	}
+	// Prepare the batch.
+	b := rt.ls.ls.Engine.NewWriteBatch()
+	defer b.Close()
+	// Update the in-memory state ahead of the truncation, so that these entries
+	// can no longer be read.
+	rt.raftMu.Lock()
+	defer rt.raftMu.Unlock()
+	require.NoError(rt.t, handleTruncatedStateBelowRaftPreApply(
+		ctx, prev, next, rt.ls.ls.StateLoader, b))
+	rt.ls.stagePendingTruncationRaftMuLocked(pendingTruncation{
+		RaftTruncatedState: next,
+		expectedFirstIndex: prev.Index + 1,
+	})
+	// Write the truncation to both storages.
+	require.NoError(rt.t, b.Commit(false /* sync */))
+	require.NoError(rt.t, rt.ms.Compact(uint64(next.Index)))
+}
+
+func (rt *replicaLogStorageTest) read(rng *rand.Rand) {
+	rt.mu.Lock()
+	defer rt.mu.Unlock()
+	// The readable interval is (compacted, readable].
+	prev, last := uint64(rt.ls.shMu.trunc.Index), rt.mu.readable
+	if last <= prev { // all compacted
+		return
+	}
+	// Bias reads to the left or right end where writes are hot.
+	if count := rng.Uint64N(last - prev); rng.IntN(2) == 0 {
+		prev += count
+	} else {
+		last -= count
+	}
+	// The moment of truth.
+	ref, err := rt.ms.Entries(prev+1, last+1, math.MaxUint64)
+	require.NoErrorf(rt.t, err, "(%d,%d]", prev, last)
+	entries, err := rt.ls.Entries(prev+1, last+1, math.MaxUint64)
+	require.NoErrorf(rt.t, err, "(%d,%d]", prev, last)
+	require.Equalf(rt.t, ref, entries, "(%d,%d]", prev, last)
+}
+
+func (rt *replicaLogStorageTest) genAppend(rng *rand.Rand) raft.StorageAppend {
+	index := rt.ms.LastIndex()
+	rt.mu.RLock()
+	committed := rt.mu.committed
+	rt.mu.RUnlock()
+	app := raft.StorageAppend{Responses: []raftpb.Message{{}}} // force sync
+
+	// Usually, raft appends entries at the end of the log. It may append with an
+	// overlap when there is a new leader. In this case, it can only append after
+	// some entry in the [committed, last] interval.
+	const pTermBump = 10
+	if termBump := rng.IntN(100) < pTermBump; termBump {
+		// NB: the committed index is >= compacted index, so the entry at the given
+		// index exists in the log.
+		index = committed + rng.Uint64N(index-committed+1)
+		rt.term++
+		app.HardState.Term = rt.term // update HardState
+	}
+	app.LeadTerm = rt.term
+	app.HardState.Commit = committed + rng.Uint64N(index-committed+1) // commit some
+
+	entries := make([]raftpb.Entry, rng.IntN(10))
+	for i := range entries {
+		entries[i] = raftpb.Entry{Index: index + 1 + uint64(i), Term: rt.term}
+	}
+	app.Entries = entries
+	return app
+}
+
+func (rt *replicaLogStorageTest) genTruncate(
+	rng *rand.Rand,
+) (prev, next kvserverpb.RaftTruncatedState) {
+	// We can only truncate entries <= committed index.
+	rt.mu.RLock()
+	committed := rt.mu.committed
+	rt.mu.RUnlock()
+	comp := rt.ms.Compacted()
```

Comment:
Should we also pull all the information from the other storage and assert that they're equal?